File: rust/nproxy/lib/src/output.rs
use crate::error::NproxyError;
use serde::Serialize;

#[derive(Serialize)]
pub struct ErrorOutput {
    pub error: ErrorDetail,
}

#[derive(Serialize)]
pub struct ErrorDetail {
    pub code: String,
    pub message: String,
}

pub fn print_yaml<T: Serialize>(data: &T) -> Result<(), NproxyError> {
    let yaml = serde_yaml::to_string(data)
        .map_err(|e| NproxyError::ApiError(format!("Failed to serialize: {}", e)))?;
    print!("{}", yaml);
    Ok(())
}

pub fn print_error(err: &NproxyError) {
    let output = ErrorOutput {
        error: ErrorDetail {
            code: err.code().to_string(),
            message: err.to_string(),
        },
    };
    if let Ok(yaml) = serde_yaml::to_string(&output) {
        eprint!("{}", yaml);
    } else {
        eprintln!("error: {}", err);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(Serialize)]
    struct TestStruct {
        name: String,
        value: i32,
    }

    #[test]
    fn test_yaml_serialization_format() {
        let data = TestStruct { name: "test".into(), value: 42 };
        let yaml = serde_yaml::to_string(&data).unwrap();
        assert!(yaml.contains("name: test"));
        assert!(yaml.contains("value: 42"));
    }

    #[test]
    fn test_error_output_format() {
        let err = NproxyError::ConfigError("test message".into());
        let output = ErrorOutput {
            error: ErrorDetail {
                code: err.code().to_string(),
                message: err.to_string(),
            },
        };
        let yaml = serde_yaml::to_string(&output).unwrap();
        assert!(yaml.contains("code: CONFIG_ERROR"));
        assert!(yaml.contains("message:"));
    }
}

File: radarr/pkg/radarr/output.go
package radarr

import (
	"fmt"
	"os"

	"gopkg.in/yaml.v3"
)

type ErrorOutput struct {
	Error ErrorDetail `yaml:"error"`
}

type ErrorDetail struct {
	Code    string `yaml:"code"`
	Message string `yaml:"message"`
}

func PrintYAML(data interface{}) error {
	enc := yaml.NewEncoder(os.Stdout)
	enc.SetIndent(2)
	if err := enc.Encode(data); err != nil {
		return APIError(fmt.Sprintf("failed to serialize: %s", err))
	}
	return nil
}

func PrintError(err error) {
	re, ok := err.(*RadarrError)
	if !ok {
		fmt.Fprintf(os.Stderr, "error: %s\n", err)
		return
	}

	output := ErrorOutput{
		Error: ErrorDetail{
			Code:    string(re.Code),
			Message: re.Message,
		},
	}

	enc := yaml.NewEncoder(os.Stderr)
	enc.SetIndent(2)
	if err := enc.Encode(output); err != nil {
		fmt.Fprintf(os.Stderr, "error: %s\n", re.Message)
	}
}
